feat: add cookies, localStorage and sessionStorage

thepushkarp opened the pull request:
Hi, while working `browser-tools-mcp`, I felt that having access to cookies, localstorage, and sessionstorage would aid the agent in debugging errors.

@tedjames let me know your thoughts and if I should make any further changes to this.

Thanks!

## Overview of Changes
- Adds cookies, localStorage, and sessionStorage to the MCP tools
- Adds Sensitive Data Settings to the Dev Tools Panel to control data security of cookies and local/session storage
- Adds an `.editorconfig` file to maintain formatting consistency

## MCP Functions introduced
- `mcp_getCookies` - Get cookies from the current page
- `mcp_getLocalStorage` - Get localStorage data
- `mcp_getSessionStorage` - Get sessionStorage data

---

This comment was updated to reflect the latest changes introduced in this PR based on further discussions.

thepushkarp committed:
add editorconfig for consistent code formatting

thepushkarp committed:
feat: Enable access to browser storage (cookies, localStorage, sessionStorage)

- Add the ability to retrieve and analyze cookies, localStorage, and sessionStorage.
- Implement new MCP tools and API endpoints for accessing browser storage data.
- Enhance the Chrome extension to capture browser storage.

tedjames commented:
Nice! I was thinking of integrating this as well!! I intentionally did not do this so that users would never send sensitive tokens to an LLM. I'm thinking we could leave this disabled by default and allow it to be configurable in the dev tools panel?

It would also be nice to persist this config to the browser...

On another note, what is everyone's thoughts on the security model for this?

I'm wondering if users opt-in for this, maybe there's a way to "detect" a token in the cookies, local storage and session storage to avoid sending. Then make that a configurable option as well ("include tokens").

Perhaps regex could be a good starting point?

thepushkarp commented:
> I'm thinking we could leave this disabled by default and allow it to be configurable in the dev tools panel?

This sounds great! Adding to it, we can have three modes to handle these data in the mcp server, all configurable from the dev tools panel:
- no access to cookies and local/session storage tools [default]
- access to cookie and local/storage tools but send no tokens — tokens can be detected and filtered out using a mix of regex (for popular tokens) and [entropy](https://blog.gitguardian.com/secrets-in-source-code-episode-3-3-building-reliable-secrets-detection) techniques
- access to everything — need to add a disclaimer that enabling this would expose your keys to the llms models

Let me know your thoughts on this :+1:

thepushkarp committed:
fix: update codebase to be more consistent with staging branch

thepushkarp committed:
fix: remove duplicate code and remove unnecessary formatting fixes

thepushkarp committed:
feat: Add sensitive data filtering and aggressive checking

- Introduces settings to control the inclusion of sensitive data.
- Implements filtering for cookies, local storage, and session storage.
- Adds entropy calculation for more aggressive sensitive data checking.

thepushkarp committed:
feat: Improve sensitive data filtering

thepushkarp committed:
fix: remove extra sensitive parameter from the cookies response

thepushkarp commented:
Hi,

The server and Chrome extension now supports the three data sensitivity settings:
- Hide All Cookies and Storage (Maximum privacy, no data will be sent to the server) -> **default**
- Hide Only Sensitive Data (Sensitive cookies and storage will be filtered out, but other cookies and storage will be sent)
- Hide Nothing (All data will be sent to the server, including potentially sensitive information)

The "Hide Only Sensitive Data" uses a mix of regex matching for all cookie and storage values along with filtering based on Shannon entropy. On smaller strings, the entropy method might lead to false-positives on some shorter strings, so it is restricted to strings longer than 16 characters.

## Screenshots

### Dev Tools Panel

![image](https://github.com/user-attachments/assets/0be1d834-43f8-417d-b3f9-7122f594d509)

### `getCookies` with "Hide All Cookies and Storage" checked

![image](https://github.com/user-attachments/assets/f0e1d503-8cbd-451d-9b89-df823856ffe7)

### `getCookies` with "Hide Only Sensitive Data" checked

![image](https://github.com/user-attachments/assets/181861c0-c3c7-4bf3-ba9e-6acc0c90491f)

### `getCookies` with "Hide Nothing" checked

![image](https://github.com/user-attachments/assets/a6d332d2-d892-430f-b2f8-a1a780d65d88)

tedjames commented:
Wow! Incredible work!! 🙏

I'm pushing out a different release now which has taken some time to prepare. Lets get this in for the next release (maybe this weekend?)

If you don't mind, can you resolve any conflicts we have with staging branch today? Happy to tackle them as well later this week - traveling a bunch right now and it's been difficult to keep up with all the updates/PRs 😅

tedjames committed:
Merge pull request #69 from AgentDeskAI/staging

Release v1.2.0

tedjames committed:
Update README.md

tedjames committed:
Update README.md

tedjames committed:
Update README.md

thepushkarp committed:
Merge branch 'staging' into add-storage

thepushkarp commented:
>  Lets get this in for the next release (maybe this weekend?)

Sure thing. I have merged the latest staging branch and have resolved the merge conflicts so we're good on that regard.
Happy travelling!

eltociear committed:
docs: update README.md

audting -> auditing

tedjames committed:
Merge pull request #76 from eltociear/patch-1

docs: update README.md

tedjames commented:
Nice!!! Amazing :)

Will review within the few days and get this in for a release next Monday at the latest if all goes well!

tedjames committed:
Update README.md

daibogh committed:
chore: update .gitignore

tedjames committed:
Merge pull request #84 from daibogh/chore/update-gitignore

chore: update .gitignore

tedjames committed:
Update README.md

thepushkarp committed:
Merge branch 'AgentDeskAI:main' into add-storage

thepushkarp commented:
hey @tedjames did you get time to review this feature? lmk if you think we need to change something

HenkDz commented:
> Nice!!! Amazing :)
> 
> Will review within the next few days and get this in for a release next Monday at the latest if all goes well!

Are you still around?
